Make sync-experiment Tiled-based

nmaytan opened the pull request:
Summary of changes so far:

- Keys in Redis are now always prefixed, even if there is no endstation ("fixes" greedy key ingestion that happens otherwise)
- The beamline and endstation acronyms are taken from the system environment, so that they don't have to be provided (but are overridable, as this won't work in some cases such as SST/SST1/SST2)
- Added facility parameter throughout to set us up for future LBMS support (still needs other work to actually function)
---
- Multiple proposals can now be activated at the same time, to support cases like at SMI where they interleave experiments. This is put in a new key `md['data_sessions_authorized']`
- Now that multiple proposals can be activated, we can enable switching proposals in a way that doesn't bypass auth as extremely -> `switch_proposal` let's you rotate through any proposals that are currently activated
- Adds the ability to switch between authorized proposals from the CLI
---
- Auth is totally Tiled-based, no more LDAP calls. This is done using `password_grant` only for now, but could use any auth flow that Tiled supports (sets us up for the future, i.e. OIDC)
- An API key is generated when `sync-experiment` runs, reducing the need to have multiple logins (API key will be used for subsequent Tiled clients)
- ~API keys are cached in Redis and managed with a priority queue based on expiration date. Keys are set to expire after 7 days.~
- ~5 keys can be cached at a time - each combination of username + proposal list counts as a new key~
- ~Only one API key is left unencrypted at a time, and it is put in Redis under a specific key so that it can be retrieved from other applications without decrypting it~
- ~The rest of the cached keys remain encrypted in Redis until they are needed~
- UPDATE: caching of API keys has been removed. Now, a single API key is stored in Redis unencrypted.
- Syncing or unsyncing the experiment now revokes the currently active API key (if one is found). If syncing, it is replaced with a new key.
- Generated keys are set to expire after 7 days.
---
- Adds the ability to "unsync" (logout) of an experiment

nmaytan committed:
Draft of Tiled-based sync-experiment

nmaytan committed:
Cleanups

nmaytan committed:
Smoothing out wrinkles

nmaytan committed:
Update requirements.txt

genematx commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,656 @@
     )
+    parser.add_argument(
+        "-s",
+        "--select",
```
should this be "activate", maybe?

genematx commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,656 @@
+                f"A proposal ID must be a 6 character integer."
+            )
+    if select_proposal not in proposal_ids:
+        raise ValueError("Cannot select a proposal which is not being activated.")
```
I think we might better stick to either "select" or "activate" and use it consistently to avoid confusion. (same as comment below)
Or are they different concepts, actually?

nmaytan replied:
Thanks for this feedback Eugene. As we discussed at the beamline, I agree with you that better words would be:
activated -> authenticated (you "authenticate" proposals for the beamline)
selected -> activated (you "activate" a specific proposal for an experiment)

I'll include this in the next round of refinement.

nmaytan committed:
Add catalog root node tags to API keys

nmaytan committed:
Upper the beamline in the root node tag

nmaytan committed:
Treat decrypt failure as missing key

nmaytan committed:
Fix vars in cache PQ

nmaytan committed:
Update terminology used by sync-experiment

Experiments (proposals) are "authorized". Then, one of the authorized
experiments is "activated". Multiple proposals may be authorized at the
same time, but only one proposal can be active at a time.

This should make more clear what the tool is doing, and how the policy
works, as opposed to before (authorize/activate vs activate/select or
activate/load).

Includes some other minor word fixes.

nmaytan committed:
Add note about duo prompt to login message

nmaytan committed:
Add unsync_experiment, update switch_proposal, correct use cases of normalized_beamline

nmaytan committed:
Sort data_sessions used for Redis keys to ensure determinism

nmaytan committed:
Make switch_proposal work from CLI

nmaytan committed:
Fix method name for authz check

genematx commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,800 @@
+        data_session.replace("-", "") for data_session in data_sessions
+    )
+    redis_prefix = f"{beamline}-{endstation}" if endstation else f"{beamline}"
+    redis_prefix = (
```
Does this mean that if a user authenticates to `pass-1` and `pass-2`, they get one api key, but later if they authenticate only to `pass-1` (or any other variation, really) a new key would need to be generated? Although, I see how checking subsets of proposals can quickly become complicated, but I wonder if there's a better way to track this.

nmaytan commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,800 @@
+    beamline = beamline.lower()
+    if endstation:
+        endstation = endstation.lower()
+    normalized_beamline = normalized_beamlines.get(beamline.lower(), beamline.lower())
```
Thanks! Small oversight on my part of this redundancy. I did the same in other places that the normalization happens, so I will fix this in multiple places.

nmaytan commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,800 @@
+
+
+facility_api_client = httpx.Client(base_url="https://api.nsls2.bnl.gov")
+normalized_beamlines = {
```
Interesting idea, I agree that'd be nice. I'll give it some thought. I think I would lean towards some implementation of defaultdict, just to be "aggressive" that this should be a simple remapping of acronyms.

nmaytan commented on nslsii/sync_experiment/sync_experiment.py:
```diff
@@ -1,292 +1,800 @@
+
+    """
+    data_sessions_sanitized = sorted(
+        data_session.replace("-", "") for data_session in data_sessions
```
While it is true that for us, a data session = `pass-` + proposal number, this is not strictly true. Data sessions could in theory have different identifiers (even possibly disconnected from the literal proposal number). I made this specific to the data session in the spirit of that possibility.

nmaytan committed:
Remove redundant lowers for beamline

nmaytan committed:
Add 'ts' for timestamp keys

nmaytan committed:
Use __missing__ to default normalized beamline

nmaytan committed:
Add prompt for Duo passcode or method

nmaytan committed:
Minor exception tweaks

nmaytan committed:
Update docstrings

nmaytan committed:
Fix access_tags dtype, update prints

nmaytan committed:
Restrict scopes on generated apikeys

nmaytan committed:
Remove apikeys scope

nmaytan committed:
Revoke api keys instead of caching

nmaytan committed:
Change from root_node tag to public tag
